jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/HttpReceiver.java

```diff
@@ -385,34 +383,26 @@ protected void responseSuccess(HttpExchange exchange, Runnable afterSuccessTask)
     protected void responseFailure(Throwable failure, Promise<Boolean> promise)
     {
         if (LOG.isDebugEnabled())
-            LOG.debug("Invoking responseFailure with {} on {}", failure, this);
-
-        invoker.run(() ->
+            LOG.debug("Failing with {} on {}", failure, this);
+
+        HttpExchange exchange = getHttpExchange();
```

Comment:
Given the comment/test below, it feels like this should be a `takeHttpExchange()` semantic, or at least the test below could be:
```java
if (exchange == null || !disassociate(exchange))
```

At the very least, can the comment below explain how the race of exchange becoming null just after the get is handled?

Reply:
@gregw I don't understand. 
The semantic cannot be "take", as the "take" happens later when also the request is completed.
The call `exchange.responseComplete()` is atomic, so the decision of whether do something or not is taken there.

Reply:
@sbordet It is a general pattern for a race to have code in an async environment that does non atomic test and then act.    It may well be that the exchange is atomically "taken" and nulled elsewhere, but a thread here may still have tested the exchange whilst non null and then act on it whilst it is null.

So either this is a race here, or there is some external circumstance that ensures that the `responseComplete` can never be called between the test and the action.    So we need to either fix the race or document why it is not a race.